File: python/metadata_compare.py
# metadata_compare.py
"""Motorul comparativ de metadate — unifica ffprobe (`media_inspector`),
Sony XML/rtmd (`sony_metadata`), EXIF/GPS si ID3 (`image_metadata`) intr-un
singur set de categorii ordonate. Port 1:1 al `MetadataCompare.swift`
(Mac) — vezi acolo pentru comentariile complete de design. Ruleaza la
cerere (cand userul deschide comparatia), NU la fiecare adaugare de
fisier in coada."""

import os

from media_inspector import probe, resolution_text
from sony_metadata import read as sony_read
from image_metadata import read_exif, read_id3

CATEGORY_ORDER = [
    "Fișier", "General", "Video", "Audio",
    "Imagine (EXIF)", "Audio (tag ID3)",
    "Setări captură (rtmd, primul cadru)", "Profil Cameră / Log (Sony XML)",
]


def _format_bytes(n):
    if n is None:
        return None
    n = float(n)
    for unit in ("B", "KB", "MB", "GB", "TB"):
        if n < 1024 or unit == "TB":
            return f"{n:.1f} {unit}" if unit != "B" else f"{int(n)} {unit}"
        n /= 1024
    return None


def categories_for(path):
    """Intoarce o lista ordonata de (categorie, [(label, valoare), ...]) —
    esec silentios per-sursa (Sony/EXIF/ID3 gol daca nu se aplica fisierului
    curent), la fel ca pe Mac."""
    cats = {}

    def add(category, label, value):
        if value is None or value == "":
            return
        cats.setdefault(category, []).append((label, str(value)))

    try:
        size = os.path.getsize(path)
    except OSError:
        size = None
    add("Fișier", "Nume fișier", os.path.basename(path))
    add("Fișier", "Dimensiune", _format_bytes(size))

    meta = probe(path)
    if meta:
        add("General", "Durată", f"{meta['duration']:.1f}s" if meta.get("duration") else None)
        add("Video", "Codec", (meta.get("video_codec") or "").upper() or None)
        add("Video", "Rezoluție", resolution_text(meta))
        add("Video", "Cadre/s", meta.get("frame_rate"))
        add("Video", "Bitrate", _format_bytes(meta["video_bitrate"] / 8) + "/s" if meta.get("video_bitrate") else None)
        add("Video", "Format pixeli", meta.get("pix_fmt"))
        add("Video", "Spațiu de culoare", meta.get("color_space"))
        add("Video", "Curbă de transfer", meta.get("color_transfer"))
        add("Video", "Gamut culoare", meta.get("color_primaries"))
        add("Audio", "Codec", (meta.get("audio_codec") or "").upper() or None)
        add("Audio", "Canale", meta.get("channels"))
        add("Audio", "Frecvență eșantionare", f"{meta['sample_rate']} Hz" if meta.get("sample_rate") else None)
        add("Audio", "Bitrate", _format_bytes(meta["audio_bitrate"] / 8) + "/s" if meta.get("audio_bitrate") else None)

    sony = sony_read(path)
    for label, value in sony.get("camera_profile", {}).items():
        add("Profil Cameră / Log (Sony XML)", label, value)
    for label, value in sony.get("capture_settings", {}).items():
        add("Setări captură (rtmd, primul cadru)", label, value)

    for label, value in read_exif(path).items():
        add("Imagine (EXIF)", label, value)
    for label, value in read_id3(path).items():
        add("Audio (tag ID3)", label, value)

    known = set(CATEGORY_ORDER)
    rest = sorted(k for k in cats if k not in known)
    order = [c for c in CATEGORY_ORDER if c in cats] + rest
    return [(c, cats[c]) for c in order]

File: CGConvertor/MetadataCompare.swift
import Foundation

/// Motorul comparativ de metadate — unifică ffprobe (`MediaInspector`),
/// Sony XML/rtmd (`SonyMetadataReader`), EXIF/GPS și ID3
/// (`ImageMetadataReader`/`ID3Reader`) într-un singur set de categorii
/// ordonate, pe modelul tabelului din `~/Developer/GDC_Metadata_View_Premium`
/// (index.html: `analyzeFile`/`categories`). Rulat la cerere (când userul
/// deschide comparația), NU la fiecare adăugare de fișier în coadă — a
/// citi un container ISO-BMFF întreg + EXIF + ID3 pentru fiecare fișier
/// automat ar fi cost inutil pentru fișierele niciodată comparate.
struct MetadataCategory: Identifiable {
    let id = UUID()
    let name: String
    let rows: [(label: String, value: String)]
}

/// Ordinea preferată a categoriilor — restul (dacă apar categorii
/// neprevăzute) ar veni după, alfabetic (nu e cazul azi, dar păstrăm
/// tiparul din index.html pentru consecvență).
private let ordineaCategoriilor = [
    "Fișier", "General", "Video", "Audio",
    "Imagine (EXIF)", "Audio (tag ID3)",
    "Setări captură (rtmd, primul cadru)", "Profil Cameră / Log (Sony XML)",
]

enum MetadataCompareEngine {
    static func categorii(pentru url: URL) -> [MetadataCategory] {
        var dict: [String: [(String, String)]] = [:]

        func adauga(_ categorie: String, _ label: String, _ valoare: String?) {
            guard let valoare, !valoare.isEmpty else { return }
            dict[categorie, default: []].append((label, valoare))
        }

        let dimensiune = (try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int64) ?? nil
        adauga("Fișier", "Nume fișier", url.lastPathComponent)
        adauga("Fișier", "Dimensiune", formatBytes(dimensiune))

        if let meta = MediaInspector.probe(url: url) {
            adauga("General", "Durată", meta.durataSecunde.map { String(format: "%.1fs", $0) })
            adauga("Video", "Codec", meta.codecVideo?.uppercased())
            adauga("Video", "Rezoluție", meta.rezolutieText)
            adauga("Video", "Cadre/s", meta.frameRate)
            adauga("Video", "Bitrate", meta.bitrateVideo.map { formatBytes(Int64($0 / 8)) + "/s" })
            adauga("Video", "Format pixeli", meta.pixFmt)
            adauga("Video", "Spațiu de culoare", meta.colorSpace)
            adauga("Video", "Curbă de transfer", meta.colorTransfer)
            adauga("Video", "Gamut culoare", meta.colorPrimaries)
            adauga("Audio", "Codec", meta.codecAudio?.uppercased())
            adauga("Audio", "Canale", meta.canaleAudio.map(String.init))
            adauga("Audio", "Frecvență eșantionare", meta.sampleRateAudio.map { "\($0) Hz" })
            adauga("Audio", "Bitrate", meta.bitrateAudio.map { formatBytes(Int64($0 / 8)) + "/s" })
        }

        let sony = SonyMetadataReader.read(from: url)
        for (label, value) in sony.cameraProfile { adauga("Profil Cameră / Log (Sony XML)", label, value) }
        for (label, value) in sony.captureSettings { adauga("Setări captură (rtmd, primul cadru)", label, value) }

        for (label, value) in ImageMetadataReader.read(from: url) { adauga("Imagine (EXIF)", label, value) }
        for (label, value) in ID3Reader.read(from: url) { adauga("Audio (tag ID3)", label, value) }

        let numeCunoscute = Set(ordineaCategoriilor)
        let restul = dict.keys.filter { !numeCunoscute.contains($0) }.sorted()
        let ordineFinala = ordineaCategoriilor.filter { dict[$0] != nil } + restul

        return ordineFinala.map { MetadataCategory(name: $0, rows: dict[$0] ?? []) }
    }
}
